### PR TITLE
Unit tests for Ebuilds

### PR DESCRIPTION
**Pull request by allenh1**
This PR adds tests for the ebuilds that are generated.

connects to #72

### CODE REVIEW

File: tests/test_ebuild.py [hunk 1]
```diff
@@ -26,7 +26,31 @@ def test_simple(self):
...
+        ebuild.description = 'an ebuild'
+        ebuild.src_uri = 'https://www.website.com/download/stuff.tar.gz'
+        ebuild.distro = 'lunar'
+        ebuild.add_run_depend('p2os_driver')
+        ebuild.add_build_depend('fake_package', False)
+        with self.assertRaises(UnresolvedDependency):
+            ebuild_text = ebuild.get_ebuild_text('Open Source Robotics Foundation', 'BSD')
+
+    def test_bad_run_depend(self):
+        """Test Bad Build Dependency"""
```
Comment (tfoote) [inline tests/test_ebuild.py]: s/build/run/

Comment (allenh1) [reply tests/test_ebuild.py]: Ah! You've caught me with my copypasta! Fixed.

**Comment by allenh1** [comment]
@tfoote I think the CI is working now if you wanna give this a look.

File: tests/test_ebuild.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    def test_external_run_depend(self):
+        """Test External Run Dependency"""
+        ebuild = Ebuild()
+        ebuild.homepage = 'https://www.website.com'
+        ebuild.description = 'an ebuild'
+        ebuild.src_uri = 'https://www.website.com/download/stuff.tar.gz'
+        ebuild.distro = 'lunar'
+        ebuild.add_run_depend('p2os_driver')
+        ebuild.add_run_depend('cmake', False)
+        ebuild_text = ebuild.get_ebuild_text('Open Source Robotics Foundation', 'BSD')
```
[the remark this answers is not in the document]
Comment (allenh1) [reply tests/test_ebuild.py]: Ya, good call.

File: tests/test_ebuild.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+class TestEbuildOutput(unittest.TestCase):
+    def test_simple(self):
+        """Test Ebuild Format"""
+        ebuild = Ebuild()
+        ebuild.homepage = 'https://www.website.com'
+        ebuild.description = 'an ebuild'
+        ebuild.add_run_depend('p2os_driver')
+        ebuild.src_uri = 'https://www.website.com/download/stuff.tar.gz'
+        ebuild.distro = 'lunar'
```
Comment (tfoote) [inline tests/test_ebuild.py]: The above 6 lines look like boilerplate that could easily be filled into a helper function to avoid repeating them below.

**Comment by allenh1** [comment]
That's a lot better, I think.

```
allenh1@hunter-laptop ~/CS/superflore
$ coverage run --source=superflore -m nose && coverage report -m
...................
----------------------------------------------------------------------
Ran 19 tests in 1.386s

OK
Name                                               Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------
superflore/TempfileManager.py                         34     23    32%   27-28, 31-41, 44-57
superflore/__init__.py                                13      4    69%   7-10
superflore/docker.py                                  36     23    36%   22-27, 30-32, 35, 38, 41-56, 61, 66
superflore/exceptions.py                              12      3    75%   18, 28, 33
superflore/generate_installers.py                     52     44    15%   31-81
superflore/generators/__init__.py                      0      0   100%
superflore/generators/bitbake/__init__.py              3      1    67%   3
superflore/generators/bitbake/gen_packages.py        115     92    20%   40-103, 109-163, 168-176, 181
superflore/generators/bitbake/ros_meta.py             27     18    33%   24-27, 30-35, 38-51, 54-55
superflore/generators/bitbake/run.py                  89     72    19%   37-42, 49-158
superflore/generators/bitbake/yocto_recipe.py        115     93    19%   41-48, 53-69, 73, 76-80, 83-95, 98-102, 105-107, 110-111, 119-122, 131-189
superflore/generators/ebuild/__init__.py               3      1    67%   3
superflore/generators/ebuild/ebuild.py               159     45    72%   120-123, 134-136, 140-160, 166-169, 180-181, 192, 205, 211-217, 222-228, 231, 241
superflore/generators/ebuild/gen_packages.py         154    127    18%   46-114, 119-151, 156-214, 219-236, 239, 242
superflore/generators/ebuild/metadata_xml.py          33     30     9%   18-24, 27-49
superflore/generators/ebuild/overlay_instance.py      36     25    31%   26-31, 34-44, 47-63, 66-67
superflore/generators/ebuild/run.py                  153    133    13%   40-43, 47-52, 59-246
superflore/repo_instance.py                           55     36    35%   26-35, 38-46, 49-57, 63, 69, 75, 81, 84-87, 92, 97
superflore/rosdep_support.py                          34      2    94%   84-85
superflore/utils.py                                  145     66    54%   32-38, 42, 46, 50, 54, 58-64, 68-73, 80, 88-98, 136, 140, 146, 151, 153-156, 158, 164, 170-171, 179, 183-184, 188-207
--------------------------------------------------------------------------------
TOTAL                                               1268    838    34%
```

The majority of uncovered territory is in the actual output, however, which is not sufficient (IMO). 

The outstanding blocks are things like [this one](https://github.com/ros-infrastructure/superflore/blob/369e292362a98f33428abc6c351b665660d9ef93/superflore/generators/ebuild/ebuild.py#L118-L123) where I checked (in what appears to be the most horrific way I could have) if something is a string or a unicode string.

@tfoote Should I just go ahead and clean up that logic here, or should that be a separate PR? Also, how do you recommend I clean up that logic?

File: tests/test_ebuild.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+# limitations under the License.
+
+from superflore.generators.ebuild.ebuild import Ebuild
+from superflore.generators.ebuild.ebuild import ebuild_keyword
+from superflore.exceptions import UnresolvedDependency
+import unittest
+
+
+class TestEbuildOutput(unittest.TestCase):
+    @classmethod
```
Comment (tfoote) [inline tests/test_ebuild.py]: You don't want classmethod here. That's for things like additional constructors. https://stackoverflow.com/questions/12179271/meaning-of-classmethod-and-staticmethod-for-beginner not just a regular method inside a class.https://docs.python.org/3/library/functions.html#classmethod

**Comment by tfoote** [comment]
There will always be more coverage to add. I see some low hanging fruit near the bottom of the ebuild with uncovered if statements that could easily be tested. 

I'd suggest looking at the unicode more hollistically. In particular you should pick a mode and then stick with it within the code. And then make sure to sanitize/homogenize any data on input so it's all of the same type. Also, it could be made as a decision to only support python3 which would drop a lot of the cases.

I have one comment on the classmethod then I suggest merging this and focusing on cleanup/coverage in a new PR/thread.

File: tests/test_ebuild.py (hunk 4 above)
Comment (allenh1) [reply tests/test_ebuild.py]: That makes sense. I was afraid that was some `unittest`-induced magic. With that definition, the context in which I had previously seen it makes more sense now.

**Comment by allenh1** [comment]
```
allenh1@hunter-laptop ~/CS/superflore
$ coverage run --source=superflore -m nose && coverage report -m            
..........................
----------------------------------------------------------------------
Ran 26 tests in 1.405s

OK
Name                                               Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------
superflore/TempfileManager.py                         34     23    32%   27-28, 31-41, 44-57
superflore/__init__.py                                13      4    69%   7-10
superflore/docker.py                                  37     23    38%   24-29, 32-34, 37, 40, 43-58, 63, 68
superflore/exceptions.py                              12      3    75%   18, 28, 33
superflore/generate_installers.py                     52     44    15%   31-81
superflore/generators/__init__.py                      0      0   100%
superflore/generators/bitbake/__init__.py              3      1    67%   3
superflore/generators/bitbake/gen_packages.py        115     92    20%   40-103, 109-163, 168-176, 181
superflore/generators/bitbake/ros_meta.py             27     18    33%   24-27, 30-35, 38-51, 54-55
superflore/generators/bitbake/run.py                  89     72    19%   37-42, 49-158
superflore/generators/bitbake/yocto_recipe.py        115     93    19%   41-48, 53-69, 73, 76-80, 83-95, 98-102, 105-107, 110-111, 119-122, 131-189
superflore/generators/ebuild/__init__.py               3      1    67%   3
superflore/generators/ebuild/ebuild.py               150     21    86%   119-122, 133-135, 139-159
superflore/generators/ebuild/gen_packages.py         154    127    18%   46-114, 119-151, 156-214, 219-236, 239, 242
superflore/generators/ebuild/metadata_xml.py          33     30     9%   18-24, 27-49
superflore/generators/ebuild/overlay_instance.py      38     26    32%   27-32, 35-45, 48-65, 68-69
superflore/generators/ebuild/run.py                  153    133    13%   40-43, 47-52, 59-246
superflore/repo_instance.py                           55     36    35%   26-35, 38-46, 49-57, 63, 69, 75, 81, 84-87, 92, 97
superflore/rosdep_support.py                          34      2    94%   84-85
superflore/utils.py                                  145     66    54%   32-38, 42, 46, 50, 54, 58-64, 68-73, 80, 88-98, 136, 140, 146, 151, 153-156, 158, 164, 170-171, 179, 183-184, 188-207
--------------------------------------------------------------------------------
TOTAL                                               1262    815    35%
```

86% percent coverage is pretty decent! I think I got all the low-hanging fruit here.

@tfoote I'll remove Python 2 in a separate commit. That should handle most of the untested code.

**Pull request merged**